### PR TITLE
TL-381: Create POST processes endpoint

### PR DESCRIPTION
**Pull request by notatiyyah**
## Link to JIRA ticket

https://hackney.atlassian.net/browse/TL-381

## Describe this PR

Create an endpoint to create new processes. This is a stub endpoint, so will improve in the future once more logic is implemented.

See SwaggerHub definition [here](https://app.swaggerhub.com/apis-docs/Hackney/ProcessesApi/1.0.0#/default/post_api_v1_process__process_name__).

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

File: ProcessesApi/V1/Factories/CreateRequestFactory.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+                    StateName = null,
+                    PermittedTriggers = new List<string>(),
+                    Assignment = new Assignment(),
+                    ProcessData = new ProcessData
+                    {
+                        FormData = createProcessQuery.FormData,
+                        Documents = createProcessQuery.Documents
+                    },
+                    CreatedAt = null,
+                    UpdatedAt = null
```
Comment (mikemoody-amido) [inline ProcessesApi/V1/Factories/CreateRequestFactory.cs:31]: Theses should either be set to the current date time here or probably better put in the gateway method. They can then be validated in the appropriate tests, wherever they get set.

File: ProcessesApi/V1/Factories/CreateRequestFactory.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
+using ProcessesApi.V1.Domain;
+using ProcessesApi.V1.Infrastructure;
+using ProcessesApi.V1.Boundary.Request;
+using System;
+using System.Collections.Generic;
+
+namespace ProcessesApi.V1.Factories
+{
+    public static class CreateRequestFactory
```
Comment (mikemoody-amido) [inline ProcessesApi/V1/Factories/CreateRequestFactory.cs:9]: Naming seems a bit off, it's not creating a request but a db entity object.

File: ProcessesApi/V1/Controllers/ProcessesApiController.cs [hunk 3]
```diff
@@ -50,5 +53,23 @@ public async Task<IActionResult> GetProcessById([FromRoute] ProcessesQuery query
...
+        /// </summary>
+        /// <response code="201">Process has been created successfully</response>
+        /// <response code="400">Bad Request</response> 
+        /// <response code="500">Internal Server Error</response>
+        [ProducesResponseType(typeof(ProcessesResponse), StatusCodes.Status201Created)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        [HttpPost]
+        [LogCall(LogLevel.Information)]
+        [Route("{process-name}")]
```
Comment (mikemoody-amido) [inline ProcessesApi/V1/Controllers/ProcessesApiController.cs:68]: Are we not using the value on the route, but only the one in the request body?

File: ProcessesApi/V1/Boundary/Request/Validation/CreateProcessQueryValidator.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+{
+    public class CreateProcessQueryValidator : AbstractValidator<CreateProcessQuery>
+    {
+        public CreateProcessQueryValidator()
+        {
+            RuleFor(x => x.TargetId).NotNull()
+                              .NotEqual(Guid.Empty);
+            RuleForEach(x => x.RelatedEntities).NotNull()
+                                        .NotEqual(Guid.Empty);
+            RuleFor(x => x.ProcessName).NotXssString()
```
Comment (mikemoody-amido) [inline ProcessesApi/V1/Boundary/Request/Validation/CreateProcessQueryValidator.cs:17]: Question: Shouldn't the ProcessName have a value (i.e. not null and not empty)?

File: ProcessesApi.Tests/V1/Gateways/ProcessesGatewayTest.cs [hunk 5]
```diff
@@ -94,5 +95,20 @@ public void GetProcessByIdExceptionIsThrown()
...
+        [Fact]
+        public async Task CreateNewProcessSucessfullySavesProcess()
+        {
+            // Arrange
+            var query = _fixture.Create<CreateProcessQuery>();
+            // Act
+            var process = await _classUnderTest.CreateNewProcess(query).ConfigureAwait(false);
+            // Assert
+            var processDb = await _dynamoDb.LoadAsync<ProcessesDb>(process.Id).ConfigureAwait(false);
+            processDb.Should().BeEquivalentTo(query.ToDatabase(), config => config.Excluding(y => y.VersionNumber));
```
Comment (mikemoody-amido) [inline ProcessesApi.Tests/V1/Gateways/ProcessesGatewayTest.cs:108]: Validate createdAt and updatedAt values...

File: ProcessesApi.Tests/V1/E2ETests/CreateNewProcessEndToEndTests.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+using System.Text;
+
+namespace ProcessesApi.Tests.V1.E2ETests
+{
+    [Collection("DynamoDb collection")]
+    public class CreateNewProcessEndToEndTests : IDisposable
+    {
+
+        private readonly Fixture _fixture = new Fixture();
+        public ProcessesDb Technology { get; private set; }
```
Comment (mikemoody-amido) [inline ProcessesApi.Tests/V1/E2ETests/CreateNewProcessEndToEndTests.cs:23]: Weird name. Also, is it used?

File: ProcessesApi/V1/Factories/CreateRequestFactory.cs (hunk 2 above)
Comment (notatiyyah) [reply ProcessesApi/V1/Factories/CreateRequestFactory.cs:9]: True, but the other APIs (e.g. Person, Tenure) use the same name. I could put it in the `EntityFactory` class?

File: ProcessesApi.Tests/V1/E2ETests/CreateNewProcessEndToEndTests.cs (hunk 6 above)
Comment (notatiyyah) [reply ProcessesApi.Tests/V1/E2ETests/CreateNewProcessEndToEndTests.cs]: No, it is not.

File: ProcessesApi/V1/Boundary/Request/Validation/CreateProcessQueryValidator.cs (hunk 4 above)
Comment (notatiyyah) [reply ProcessesApi/V1/Boundary/Request/Validation/CreateProcessQueryValidator.cs]: Probably..

File: ProcessesApi/V1/Controllers/ProcessesApiController.cs (hunk 3 above)
Comment (notatiyyah) [reply ProcessesApi/V1/Controllers/ProcessesApiController.cs]: Yes, we are. I thought that was how it was done, but should be fixed now...?

**Review by mikemoody-amido — APPROVED** [review]

**Pull request merged**